### PR TITLE
8331185: Enable compiler memory limits in debug builds

### PR DESCRIPTION
**Pull request by tstuefe**
See [1] for previous discussions.

We'd like to introduce a default memory limit for compilations in debug builds. That way, we can catch pathological compiler errors that have an unreasonably high per-compilation memory footprint early during testing.

The default limit affects all compilations, unless the method is subject to a memory limit set from command line. Meaning, `-XX:CompileCommand=MemLimit,...` overrules the default.

Examples:

This lowers the memlimit for j.l.String methods - all methods will have the default 1GB limit in a debug JVM. Only j.l.String will run with a 100M limit: `-XX:CompileCommand=MemLimit,java.lang.String::*,100m`

This disables the default memlimit globally: `-XX:CompileCommand=MemLimit,*.*,0`

---

The patch:

1) adds a debug-only default memory limit of **1GB** (as proposed by @vnkozlov). The limit action is "crash", meaning we will assert.
2) To test the mechanics, we now print out the memory limit for each compilation in the compilation cost record.
3) Adapted and extended tests

I also fixed up some copyrights that I overlooked last year when adding the compiler memory statistics this patch builds atop of.

Tested:

- manually on Mac m1 (debug and release)
- GHAs are running
- but Oracle will do more testing before this goes in

[1] https://mail.openjdk.org/pipermail/hotspot-compiler-dev/2024-April/074787.html

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331185](https://bugs.openjdk.org/browse/JDK-8331185): Enable compiler memory limits in debug builds (**Enhancement** - P4)

### Reviewers
 * [Ashutosh Mehra](https://openjdk.org/census#asmehra) (@ashu-mehra - Committer) ⚠️ Review applies to [e2aacaed](https://git.openjdk.org/jdk/pull/18969/files/e2aacaed0acb689594529a0c32e3955b54b30651)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18969/head:pull/18969` \
`$ git checkout pull/18969`

Update a local copy of the PR: \
`$ git checkout pull/18969` \
`$ git pull https://git.openjdk.org/jdk.git pull/18969/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18969`

View PR using the GUI difftool: \
`$ git pr show -t 18969`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18969.diff">https://git.openjdk.org/jdk/pull/18969.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18969#issuecomment-2079213872)

### CODE REVIEW

**Comment by vnkozlov** [comment]
I submitted our testing.

Why you did not merge your latest [JDK-8330625](https://github.com/openjdk/jdk/commit/2b7176a55ad0e5c6ba34abba3fe8fc1a411a5e2d) change? Patch was applied with offsets. So I will test latest JDK.

File: test/hotspot/jtreg/compiler/print/CompileCommandPrintMemStat.java [hunk 1]
```diff
@@ -1,5 +1,6 @@
 /*
- * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2023, 2024 Red Hat, Inc. All rights reserved.
+ * Copyright (c) 2023, 2024 Oracle and/or its affiliates. All rights reserved.
```
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/print/CompileCommandPrintMemStat.java:3]: Missing comma `,` after second year. Files header verifier failed for both test files.

**Comment by vnkozlov** [comment]
`compiler/c2/TestFindNode.java` and `compiler/loopopts/TestDeepGraphVerifyIterativeGVN.java` failed on `aarch64` when run with stress flags:

`-XX:+UnlockDiagnosticVMOptions -XX:-TieredCompilation -XX:+StressArrayCopyMacroNode -XX:+StressLCM -XX:+StressGCM -XX:+StressIGVN -XX:+StressCCP -XX:+StressMacroExpansion -XX:+StressMethodHandleLinkerInlining -XX:+StressCompiledExceptionHandlers`

```
#  Internal Error (/workspace/open/src/hotspot/share/compiler/compilationMemoryStatistic.cpp:551), pid=912622, tid=912639
#  fatal error: c2 compiler/c2/TestFindNode::test(()V): Hit MemLimit (limit: 1073741824 now: 1073824544)

V  [libjvm.so+0x97496c]  report_fatal(VMErrorType, char const*, int, char const*, ...)+0x108  (debug.cpp:214)
V  [libjvm.so+0x8a9e2c]  CompilationMemoryStatistic::on_arena_change(long, Arena const*)+0x5bc  (compilationMemoryStatistic.cpp:551)
V  [libjvm.so+0x58ee70]  Arena::grow(unsigned long, AllocFailStrategy::AllocFailEnum)+0x10c  (arena.cpp:300)
V  [libjvm.so+0x13eb86c]  PhaseChaitin::Split(unsigned int, ResourceArea*)+0x3ac  (resourceArea.inline.hpp:35)
V  [libjvm.so+0x798a44]  PhaseChaitin::Register_Allocate()+0x554  (chaitin.cpp:553)
V  [libjvm.so+0x8ced88]  Compile::Code_Gen()+0x284  (compile.cpp:2988)
V  [libjvm.so+0x8d1128]  Compile::Compile(ciEnv*, ciMethod*, int, Options, DirectiveSet*)+0x1588  (compile.cpp:896)
V  [libjvm.so+0x72d480]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x17c  (c2compiler.cpp:142)
```

Looks like Register Allocator eats memory.

**Comment by vnkozlov** [comment]
There also few closed tests which failed with `-XX:-TieredCompilation -XX:+AlwaysIncrementalInline` on `x64` and `aarch64`:

```
V  [libjvm.so+0xa90b95]  report_fatal(VMErrorType, char const*, int, char const*, ...)+0x105  (debug.cpp:214)
V  [libjvm.so+0x9c06dd]  CompilationMemoryStatistic::on_arena_change(long, Arena const*)+0x5cd  (compilationMemoryStatistic.cpp:551)
V  [libjvm.so+0x5f45fc]  Arena::grow(unsigned long, AllocFailStrategy::AllocFailEnum)+0x10c  (arena.cpp:300)
V  [libjvm.so+0x14e4706]  Parse::init_blocks()+0x46  (parse1.cpp:1290)
V  [libjvm.so+0x14f0aff]  Parse::Parse(JVMState*, ciMethod*, float)+0x52f  (parse1.cpp:565)
V  [libjvm.so+0x8412b9]  ParseGenerator::generate(JVMState*)+0x169  (callGenerator.cpp:99)
V  [libjvm.so+0x84482a]  PredictedCallGenerator::generate(JVMState*)+0x3aa  (callGenerator.cpp:928)
V  [libjvm.so+0x8466de]  CallGenerator::do_late_inline_helper()+0x90e  (callGenerator.cpp:704)
V  [libjvm.so+0x9e2de4]  Compile::inline_incrementally_one()+0xd4  (compile.cpp:2054)
V  [libjvm.so+0x9e3b32]  Compile::inline_incrementally(PhaseIterGVN&)+0x292  (compile.cpp:2137)
V  [libjvm.so+0x9e5b50]  Compile::Optimize()+0x340  (compile.cpp:2272)
V  [libjvm.so+0x9e9c60]  Compile::Compile(ciEnv*, ciMethod*, int, Options, DirectiveSet*)+0x1b50  (compile.cpp:863)
V  [libjvm.so+0x83ed15]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x1d5  (c2compiler.cpp:142)
```

**Comment by vnkozlov** [comment]
Closed test passed with `1200M` limit

**Comment by tstuefe** [comment]
@vnkozlov How do you want to handle this? Are these memory usage numbers pathological or normal?

Should I increase the general limit to 1200M? Alternatively, we can also just run failing the test with a higher memory limit.

**Comment by vnkozlov** [comment]
> @vnkozlov How do you want to handle this? Are these memory usage numbers pathological or normal?
> 
> Should I increase the general limit to 1200M? Alternatively, we can also just run failing the test with a higher memory limit.

The closed failed test has very log chain of lambda forms and I don't think memory consumption is pathological. May be something could be done to improve it (as enhancement RFE) but it is not urgent. I will help you with increase limit for it.

I did not check which limit will allow to pass failed open tests. I will ask you to investigate and increase limit for them for now and file bugs to investigate the cause.

I also need to run later tiers (I ran only up to tier5) before integration. If I find more failing cases we may consider increase default limit.

**Comment by vnkozlov** [comment]
I did not find any other failures in tier6-9. So let's adjust limit and file bugs for failed tests I found before. And keep default limit 1Gb.

**Comment by tstuefe** [comment]
> I submitted our testing.
> 
> Why you did not merge your latest [JDK-8330625](https://github.com/openjdk/jdk/commit/2b7176a55ad0e5c6ba34abba3fe8fc1a411a5e2d) change? Patch was applied with offsets. So I will test latest JDK.

Ah forgot. Sorry. Thanks for testing with the latest head.

**Comment by tstuefe** [comment]
@vnkozlov Opened a follow up issue for `compiler/c2/TestFindNode.java` : https://bugs.openjdk.org/browse/JDK-8331283

**Comment by tstuefe** [comment]
@vnkozlov ... but I was unable to reproduce the problem with `compiler/loopopts/TestDeepGraphVerifyIterativeGVN.java` on aarch64. Memory usage during compilation of the test method is ~170MB, which is fine.

I tried to reproduce it on both MacOS m1, and on a Raspberry 4 with 64bit linux.

**Comment by tstuefe** [comment]
Opened a bug to track the memory limit break in compiler/loopopts/TestDeepGraphVerifyIterativeGVN.java. Since I cannot reproduce it, someone at Oracle should look at this:  https://bugs.openjdk.org/browse/JDK-8331295

**Comment by vnkozlov** [comment]
Thank you, @tstuefe, for filing these bugs.

One additional thing I noticed is that we don't produce compilation replay file (its size is 0) for such failures. Can you look why is that?

**Comment by vnkozlov** [comment]
I filed bug for our closed test failure.

File: test/hotspot/jtreg/compiler/loopopts/TestDeepGraphVerifyIterativeGVN.java [hunk 2]
```diff
@@ -27,12 +27,14 @@
...
  * @summary Test which causes a stack overflow segmentation fault with -XX:VerifyIterativeGVN=1 due to a too deep recursion in Node::verify_recur().
  *
- * @run main/othervm/timeout=600 -Xcomp -XX:VerifyIterativeGVN=1 -XX:CompileCommand=compileonly,compiler.loopopts.TestDeepGraphVerifyIterativeGVN::*
+ * @run main/othervm/timeout=600 -Xcomp -XX:VerifyIterativeGVN=1 -XX:CompileCommand=compileonly,compiler.loopopts.TestDeepGraphVerifyIterativeGVN::* -XX:CompileCommand=memlimit,TestFindNode::test,0
  *                               compiler.loopopts.TestDeepGraphVerifyIterativeGVN
  */
 
 package compiler.loopopts;
 
+// Note; we disable the implicit memory limit of 1G in debug JVMs until JDK-8331283 is fixed
```
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/loopopts/TestDeepGraphVerifyIterativeGVN.java:36]: Different bug for this test: JDK-8331295

File: test/hotspot/jtreg/compiler/loopopts/TestDeepGraphVerifyIterativeGVN.java [hunk 3]
```diff
@@ -27,12 +27,14 @@
  * @requires vm.debug == true & vm.flavor == "server"
  * @summary Test which causes a stack overflow segmentation fault with -XX:VerifyIterativeGVN=1 due to a too deep recursion in Node::verify_recur().
  *
- * @run main/othervm/timeout=600 -Xcomp -XX:VerifyIterativeGVN=1 -XX:CompileCommand=compileonly,compiler.loopopts.TestDeepGraphVerifyIterativeGVN::*
+ * @run main/othervm/timeout=600 -Xcomp -XX:VerifyIterativeGVN=1 -XX:CompileCommand=compileonly,compiler.loopopts.TestDeepGraphVerifyIterativeGVN::* -XX:CompileCommand=memlimit,TestFindNode::test,0
```
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/loopopts/TestDeepGraphVerifyIterativeGVN.java:30]: Why you list TestsFindNode here?
It is mistake in JDK-8331295 Description referencing TestsFindNode failure output. I added comment in JBS with correct TestDeepGraphVerifyIterativeGVN failure output.

**Comment by tstuefe** [comment]
> Thank you, @tstuefe, for filing these bugs.
> 
> One additional thing I noticed is that we don't produce compilation replay file (its size is 0) for such failures. Can you look why is that?

Yes, its https://bugs.openjdk.org/browse/JDK-8331344 . I'll post a PR shortly.

The problem behind this is more generic, namely that producing replay files needs resource area, and it shouldn't. We should not allocate resource area or heap in fatal error handling. But for now, I'll fix this locally by avoiding the recursion.

**Comment by vnkozlov** [comment]
> > Thank you, @tstuefe, for filing these bugs.
> > One additional thing I noticed is that we don't produce compilation replay file (its size is 0) for such failures. Can you look why is that?
> 
> Yes, its https://bugs.openjdk.org/browse/JDK-8331344 . I'll post a PR shortly.
> 
> The problem behind this is more generic, namely that producing replay files needs resource area, and it shouldn't. We should not allocate resource area or heap in fatal error handling. But for now, I'll fix this locally by avoiding the recursion.

Good. I think we need to push it before this PR.

**Comment by tstuefe** [comment]
> > > Thank you, @tstuefe, for filing these bugs.
> > > One additional thing I noticed is that we don't produce compilation replay file (its size is 0) for such failures. Can you look why is that?
> > 
> > 
> > Yes, its https://bugs.openjdk.org/browse/JDK-8331344 . I'll post a PR shortly.
> > The problem behind this is more generic, namely that producing replay files needs resource area, and it shouldn't. We should not allocate resource area or heap in fatal error handling. But for now, I'll fix this locally by avoiding the recursion.
> 
> Good. I think we need to push it before this PR.

@vnkozlov Openend PR: https://github.com/openjdk/jdk/pull/19005

File: test/hotspot/jtreg/compiler/print/CompileCommandMemLimit.java [hunk 4]
```diff
@@ -109,31 +110,47 @@ private static void test(String include, String exclude) throws Exception {
...
+            oa.shouldNotMatch(method1regex + ".*Hit MemLimit");
+            oa.shouldNotMatch(method3regex + ".*Hit MemLimit");
 
-            // Expect final output to contain "oom"
-            oa.shouldMatch(".*oom.*" + expectedNameIncl + ".*");
+            // Final report:
+            // Method 1 should show up as "ok" and with the default limit, e.g.
+            // total     NA        RA        result  #nodes  limit   time    type  #rc thread              method
+            // 32728     0         32728     ok     -       1024M   0.045   c1    1   0x000000011b019c10  compiler/print/CompileCommandMemLimit$TestMain::method1(()J)
+            oa.shouldMatch("\\d+ +\\d+ +\\d+ +ok +" + numberNodesRegex + " +" + implicitMemoryLimit + " +.* +" + method1regex);
```
Comment (ashu-mehra) [inline test/hotspot/jtreg/compiler/print/CompileCommandMemLimit.java:143]: A minor suggestion regarding the regex. I find "\s+" more readable than " +" to match multiple spaces.

**Review by ashu-mehra — APPROVED** [review]

**Comment by ashu-mehra** [comment]
Just one suggestion which you may pick or ignore. Otherwise looks good.

**Comment by tstuefe** [comment]
> Just one suggestion which you may pick or ignore. Otherwise looks good.

Many thanks, @ashu-mehra ! 

I will actually ignore your suggestion, because I want the expression to only match spaces precisely, not whitespaces. But for any-whitespace, I usually do as you suggest.

**Comment by tstuefe** [comment]
@vnkozlov SAP did a test series and did not find any issues in their CI

**Comment by vnkozlov** [comment]
Okay, I will run our testing too.

**Comment by vnkozlov** [comment]
Looks like `memlimit,TestFindNode::test,0` does not work. The test failed with stress flags [JDK-8331283](https://bugs.openjdk.org/browse/JDK-8331283) on linux-aarch64 (Ampere). With the same call stack. I see `-XX:CompileCommand=memlimit,TestFindNode::test,0` in flags passed to test.

**Comment by vnkozlov** [comment]
I attached hs_err file to [JDK-8331283](https://bugs.openjdk.org/browse/JDK-8331283)

**Comment by tstuefe** [comment]
> Looks like `memlimit,TestFindNode::test,0` does not work. The test failed with stress flags [JDK-8331283](https://bugs.openjdk.org/browse/JDK-8331283) on linux-aarch64 (Ampere). With the same call stack. I see `-XX:CompileCommand=memlimit,TestFindNode::test,0` in flags passed to test.

I fixed the error, a simple typo (forgot to properly name the class in the option). Retested locally on Mac m1, confirmed that the test passes with this commit, fails without it. I am not sure what went wrong, since I did these tests beforehand. Maybe I pushed the wrong version.

That is slightly concerning, however, since the error should have come up at SAP too. I guess they don't test with all these stress options.

**Comment by vnkozlov** [comment]
`-XX:CompileCommand=memstat,compiler.c2.TestFindNode::*,print` - leftover from debugging?

**Comment by tstuefe** [comment]
> `-XX:CompileCommand=memstat,compiler.c2.TestFindNode::*,print` - leftover from debugging?

I tend to leave debug output in, if its not too large, to speed up any follow-up fixes I need to do later. But then, I did not do it consistently anyway, so I removed the output.

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Comment by tstuefe** [comment]
Many thanks, @vnkozlov !

/integrate

**Pull request closed without merging**